src/server/services/wms/qgswmsrendercontext.h

```diff
@@ -292,7 +292,9 @@ namespace QgsWms
       void searchLayersToRenderStyle();
       void removeUnwantedLayers();
 
-      void checkLayerReadPermissions();
+      void addLayerToRender( QgsMapLayer *layer, bool queryLayer );
+
+      bool checkLayerReadPermissions( QgsMapLayer *layer, bool queryLayer );
```

Comment:
```suggestion
      /**
       * Adds the layer to the list of layers to be rendered if the layer is readable
       * Throws a QgsSecurityException if the layer is not readable and it is directly requested (queryLayer = true)
       */
      void addLayerToRender( QgsMapLayer *layer, bool queryLayer );

      /**
       * Check layer read permissions
       * Returns True if the layer is readable
       * Returns False if the layer is not readable and it is not directly requested (queryLayer = false, the layer is in a group that has been requested)
       * Throws a QgsSecurityException if the layer is not readable and it is directly requested (queryLayer = true)
       */
      bool checkLayerReadPermissions( QgsMapLayer *layer, bool queryLayer );
```
@elpaso are these comments suitable?

Reply:
That looks better, thanks. Just one thing that it is not clear to me in checkLayerReadPermissions documentation:
```Returns False if the layer is not readable and it is not directly requested (queryLayer = false, the layer is in a group that has been requested)``` the text in parenthesis is not clear, shouldn't it be ```(queryLayer = false or the layer is in a group that has been requested)```?

Reply:
@elpaso `queryLayer == false` if the layer is ina group that has been requested or the layer has not been directly requested by has to render.

Reply:
Ok, so I recommend to remove the queryLayer parameter completely and handle the throw in client code. It doesn't make sense to me to pass a bool to trigger a throw. You can return false and throw the exception in client code, or am I missing something?

Reply:
> Ok, so I recommend to remove the queryLayer parameter completely and handle the throw in client code. 

This is not enough: you need to check if the layer is allowed the add it to the group  or not without throwing exception depending on the context, so we need to create different methods to handle `queryLayer == false` cases. The construct was just convenient way to limit code verbosity.

Furthemore I was always told that it was not  a good practice to handle control flows with exceptions in C++: if the parameter is a problem, then I would rather go with different methods similar to `checkLayerPermission` and `addLayerToRender` that do not throw.

Reply:
Removed  `querLayer` param and throw exception in client code.